pkg/provider/github/github.go

```diff
@@ -0,0 +1,110 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://wwg.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package github
+
+import (
+	"context"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/golang-jwt/jwt/v5"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+)
+
+type Github struct {
+	http      *http.Client
+	kube      client.Client
+	namespace string
+	store     esv1beta1.GenericStore
+	storeKind string
+	url       string
+}
+
+func (g *Github) GetPrivateKeyAppID(ctx context.Context) (*rsa.PrivateKey, string, error) {
+	provider, err := getProvider(g.store)
+	if err != nil {
+		return nil, "", fmt.Errorf("can't get provider: %w", err)
+	}
+
+	key, err := g.getStoreSecret(ctx, provider.Auth.SecretRef.PrivatKey)
+	if err != nil {
+		return nil, "", fmt.Errorf("can't get provider auth secret: %w", err)
+	}
+	pk, err := jwt.ParseRSAPrivateKeyFromPEM(key.Data[provider.Auth.SecretRef.PrivatKey.Key])
+	if err != nil {
+		return nil, "", fmt.Errorf("error parsing RSA private key: %w", err)
+	}
+	return pk, provider.AppID, nil
+}
+
+// Get github installation token.
+func GetInstallationToken(key *rsa.PrivateKey, aid string) (string, error) {
+	claims := jwt.RegisteredClaims{
+		Issuer:    aid,
+		IssuedAt:  jwt.NewNumericDate(time.Now().Add(-time.Second * 10)),
+		ExpiresAt: jwt.NewNumericDate(time.Now().Add(time.Second * 300)),
+	}
+
+	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
+	signedToken, err := token.SignedString(key)
+	if err != nil {
+		return "", fmt.Errorf("error signing token: %w", err)
+	}
+
+	return signedToken, nil
+}
+
+func (g *Github) GetSecret(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	key, appID, err := g.GetPrivateKeyAppID(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error getting private key: %w", err)
+	}
+
+	itoken, err := GetInstallationToken(key, appID)
+	if err != nil {
+		return nil, fmt.Errorf("can't get InstallationToken: %w", err)
+	}
```

Comment:
the token signing can be done once when the provider is constructed in `NewClient()`.